chore(serverless-plugin): initial commit

mbaltrusitis opened the pull request:
<!-- Please ensure you have done the following steps: -->

- [ ] Followed the steps in the contributor's guide: https://crashoverride.com/docs/other/contributing#filing-the-pull-request
- [ ] PR title uses [semantic commit messages](https://nitayneeman.com/posts/understanding-semantic-commit-messages-using-git-and-angular/#fix)
- [ ] Filled out the template to a useful degree
- [ ] Updated `CHANGELOG.md` if necessary

## Issue

<!-- Link the ticket(s) that this PR works on. Every pr should have a linked issue. -->

## Description

Initial commit to add Serverless Framework plugin.

<!-- What does this PR do? A list of changes would be useful for larger PRs. -->

## Testing

<!-- What are the steps needed to test this PR? -->

mbaltrusitis committed:
chore(serverless-plugin): initial commit

DarbyBrown reviewed (commented):
some minor pedantry, but this all looks good to me

DarbyBrown reviewed (dismissed):
some minor pedantry, but this all looks good to me

Francisc0-F3rnand3s commented on serverless-plugin-crashoverride/README.md:
```diff
@@ -0,0 +1,132 @@
+├── src/
+│   ├── index.ts          # Main plugin class
+│   └── types.ts          # TypeScript type definitions
+├── tests/
```
I would colocating tests as this makes it easier to spot test files for the implementation.

You can use the test directory for any setup/test utils

DarbyBrown replied:
This seems common in all the severless plugins I've looked at

mbaltrusitis replied:
will do!

mbaltrusitis committed:
refactor(serverless-framework): co-locate tests

mbaltrusitis committed:
feat(serverless-framework): memorySize check

When custom.crashoverride.memoryCheck is true, if the configured
memory is <512MB, then the build will fail closed with a failure.

mbaltrusitis committed:
feat(serverless-plugin): add chalkCheck configuration to enforce chalk binary availability

  - Add chalkCheck boolean option to CrashOverrideConfig interface
  - Implement chalk binary validation in chalkBinaryAvailable() method
  - Fail build when chalkCheck=true and chalk binary is missing
  - Warn only when chalkCheck=false (default) and chalk is missing
  - Add comprehensive test coverage for chalkCheck feature

miki725 reviewed (changes requested):
should this be moved under dust extension folder? both are related so prolly good idea to keep them side-by-side

mbaltrusitis committed:
refactor: config

mbaltrusitis committed:
refactor: logging

mbaltrusitis committed:
feat(serverless-plugin): dynamically fetch ARN + docs and tests

Fetch region-appropriate ARN for dust extension. Replace tests
with a mock-based approach avoiding calls to the framework,
network and filesystem.

- After provider config has been loaded, fetch ARN and run memory check
- Add TESTING.md explainer for approach, mocks and utility functions
- Update module and function documentation where appropriate

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,276 @@
+      | CustomServerlessConfig
+      | undefined;
+    const memoryCheckEnabled =
+      customConfig?.crashoverride?.memoryCheck ?? false;
```
refactored this to use the spread `...` operatior

mbaltrusitis committed:
refactor(serverless-plugin): stricter provider typing

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,276 @@
+  public options: Serverless.Options;
+  public hooks: Plugin.Hooks;
+  public provider: any;
+  private log: any;
```
pulled in the AWS types which resolved most of these instances.

mbaltrusitis committed:
feat(serverless-plugin): add platform guard for Windows

Prevent the plugin from intializing // registering hooks to
ensure normal operation on local, Windows-based development
machines.

mbaltrusitis committed:
refactor(serverless-plugin): update error message

With a User-facing exception thrown, prompt the user with a fix

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,276 @@
+        ),
+      );
+    } catch (error: any) {
+      this.log.error(
```
now blows up. note: it will now throw `serverless.classes.Error` which is considered a "user" error over an "programming" error. The error message will prompt the user to make chalk available in `PATH`.

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,276 @@
+  public serverless: Serverless;
+  public options: Serverless.Options;
+  public hooks: Plugin.Hooks;
+  public provider: any;
```
latest push should have removed any `any`s

mbaltrusitis commented on serverless-plugin-crashoverride/package.json:
```diff
@@ -0,0 +1,43 @@
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "build": "tsc",
```
great resource, thanks for sharing. tsconfig.json was originally omitted because its in the `chalk` gitignore's *.json rule.

mbaltrusitis committed:
chore(serverless-plugin): add tsconfig.json

Derived from https://github.com/tsconfig/bases

- based on Node 18
- strict checks
- inherits `target: "es2022"` and `lib: ["es2023"] from bases

mbaltrusitis committed:
refactor(serverless-plugin): use `which` over `command`

mbaltrusitis committed:
feat(serverless-plugin): add Dust enforcement test layerCheck

Implements layer validation for AWS Lambda functions to ensure Dust extension
is properly configured across all deployed functions.

Key features:
- Validates CloudFormation templates after packaging
- Checks all Lambda functions for Dust extension ARN presence
- Configurable to fail open or close via layerCheck option
- Status reporting showing validation results

When layerCheck=true: (fail close) throws an error if any function missing extension
When layerCheck=false: (fail open) logs a warning but allows deployment to continue

mbaltrusitis committed:
refactor(serverless-plugin): test refactor

Removes all console-based assertions, replacing them
with tests on application state. As a result some
application logic has also been refactored into pure funcs
when possible. Functions registered with lifecycle hooks
are "coordinators" assembling necessary state for pure funcs
and driving logic flow based on their return values.

mbaltrusitis committed:
refactor(serverless-plugin): update package namespace

mbaltrusitis committed:
refactor(serverless-plugin): mark attribues as private and/or readonly

mbaltrusitis committed:
refactor(serverless-plugin): migrate from `interface` to `type`

Also remove use of `any`

mbaltrusitis committed:
feat(serverless-plugin): Add Dust Extension ARN version pinning

Enable users to pin the Dust extension to a specific version
using the arnVersion custom config value or, via the env
variable `CO_ARN_VERSION`. The value must be a positive
integer to pass validation.

mbaltrusitis committed:
test(serverless-plugin): do not use local hooks

- Remove test and build from pre-commit
- Migrate remaining lint and prettier hooks to prebuilt env

mbaltrusitis committed:
lint(serverless-plugin): semicolons

mbaltrusitis commented on serverless-plugin-crashoverride/src/__mocks__/https.ts:
I have a singular HTTP request in here. I figured I'd avoid the dependency.

mbaltrusitis committed:
chore(serverless-plugin): remove duplicate pre-commit config

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,385 @@
+
+    private fetchProviderConfig(): void {
+        if (this.providerConfig != null) {
+            // early exit if config has already been fetched
```
the `serverless` CLI can be invoked to start at a life cycle stage that may be after when `fetchProviderConfig` is registered. I've added these guards to be a bit more defensive. In the refactor I did though, I assert it on the "coordinator functions" that are directly registered with the life cycle to reduce the amount of calls.

mbaltrusitis committed:
feat(serverless-plugin): prevent duplicate Dust extension

If a user manually adds a Dust Extension ARN, skip adding the
Dust Extension to that particular function and report it as
skipped.

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,385 @@
+            if (!awsFunc.layers) {
+                awsFunc.layers = [];
+            }
+            awsFunc.layers = [...(awsFunc.layers ?? []), extensionArn];
```
added a test for comparing against base ARN (no version)

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,385 @@
+            if (currentLayers.length >= MAX_LAYERS_AND_EXTENSIONS) {
+                const error = `Cannot add Dust Lambda Extension to function ${chalk.bold(functionName)}: would exceed maximum layer/extension limit of ${MAX_LAYERS_AND_EXTENSIONS} (currently has ${currentLayers.length})`;
+                this.log_error(error);
+                throw new this.serverless.classes.Error(error);
```
I was intending on keeping operations atomic.

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,385 @@
+        for (const [functionName, func] of Object.entries(functions)) {
+            if (!func) return;
+
+            const awsFunc = func as Aws.AwsFunction;
```
Serverless Framework returns an agnostic Union type of `FunctionDefinitionHandler | FunctionDefinitionImage` which does not include the `layers` attribute. At runtime, when the provider value of `aws` is given, these types are then populated with AWS-specific properties. Take a peep at `types.ts` and let me know if you think that's a reasonable workaround.

miki725 commented on serverless-plugin-crashoverride/tsconfig.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "extends": ["@tsconfig/node18/tsconfig.json", "@tsconfig/strictest/tsconfig.json"],
```
serverless framework doesnt actually run in lambda runtime. its just a means to deploy a lambda function into AWS. I would recommend to use more modern node version

mbaltrusitis committed:
refactor(serverless-plugin): avoid explicit type masking

Serverless Framework treats functions as cloud-agnostic and
therefore lacks the AWS-specific attribute of a layers array.
To avoid explicit casting or the use of `any`, a type for
RuntimeAwsFunction was created to represent the hybrid nature
of a FunctionDefinitionHandler that is enriched with AWS-
specific attributes at runtime.

mbaltrusitis commented on .pre-commit-config.yaml:
```diff
@@ -29,8 +29,46 @@ repos:
+      # lint
+      - id: serverless-plugin-eslint
+        name: ESLint (serverless-plugin-crashoverride)
+        entry: bash -c "cd serverless-plugin-crashoverride && npm run lint"
```
adjusted

miki725 commented on serverless-plugin-crashoverride/src/index.ts:
```diff
@@ -0,0 +1,385 @@
+        // Persist provider configuration
+        this.providerConfig = {
+            provider: "aws", // We're specifically an AWS plugin
+            region: provider.region || "us-east-1", // AWS default region
```
https://github.com/serverless/serverless/issues/2151

🤯 

this is wildly incorrect from the serverless side...

mbaltrusitis committed:
refactor(serverless-plugin): migrate from https.get to fetch

mbaltrusitis committed:
refactor(serverless-plugin): update Node from 18 -> 22

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.test.ts:
```diff
@@ -0,0 +1,579 @@
+        expect(() => executeDeploymentHook(plugin)).toThrow(ServerlessError);
+        expect(() => executeDeploymentHook(plugin)).toThrow(
+          "Memory check failed: memorySize (1024MB) is less than minimum required (2048MB)"
+        );
```
most of the tests have been refactored to avoid this, along with the application code, keeping them as pure as possible.

mbaltrusitis commented on serverless-plugin-crashoverride/src/types.ts:
```diff
@@ -0,0 +1,15 @@
+  chalkCheck?: boolean;
+}
+
+export interface CustomServerlessConfig {
```
refactored from interfaces to types

mbaltrusitis commented on .pre-commit-config.yaml:
```diff
@@ -29,8 +29,46 @@ repos:
+      - id: serverless-plugin-eslint
+        name: ESLint (serverless-plugin-crashoverride)
+        entry: bash -c "cd serverless-plugin-crashoverride && npm run lint"
+        language: system
```
updated!

mbaltrusitis committed:
refactor(serverless-plugin): tidy-up constructor

mbaltrusitis commented on serverless-plugin-crashoverride/src/index.test.ts:
I've further refactored the tests in commit a0ac3b26525a4919afdcf9482a5ebd4bf290051a

mbaltrusitis committed:
chore(serverless-plugin): add prettier config

mbaltrusitis committed:
refactor(serverless-plugin): remove layerCheck
